raster/src/main/scala/geotrellis/raster/summary/polygonal/visitors/TileFoldingVisitor.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2019 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package geotrellis.raster.summary.polygonal.visitors
+
+import geotrellis.raster._
+import geotrellis.raster.summary.CellVisitor
+
+abstract class TileFoldingVisitor
+    extends CellVisitor[Raster[Tile], Option[Double]] {
+  private var accumulator: Option[Double] = None
+
+  def result: Option[Double] = accumulator
+
+  def visit(raster: Raster[Tile], col: Int, row: Int): Unit = {
+    val newValue = raster.tile.getDouble(col, row)
+    accumulator = result match {
+      case Some(accum) if isData(newValue) => Some(fold(accum, newValue))
```

Comment:
@echeipesh we probably need to create a task to update https://github.com/geotrellis/geotrellis-benchmark/tree/master/geotrellis/src/test/scala/geotrellis/raster/mapalgebra/focal and move into contrib(?)

Reply:
Not sure, we have `bench` project in GeoTrellis. Would you suggest keeping them in separate repo?

Reply:
We can do both; According to the README, we deprecated this project, since it uses `scaliper` that is deprecated and not maintained. The replacement is a bench project in geotrellis-contrib (again, according to the README).

Reply:
```
[info] Benchmark                            Mode  Cnt  Score   Error  Units
[info] PolygonalSummaryBench.cleanVisitor   avgt    3  0.075 ± 0.045   s/op
[info] PolygonalSummaryBench.frugalVisitor  avgt    3  0.017 ± 0.022   s/op
```

Reply:
https://github.com/echeipesh/geotrellis/tree/bench/polygonal-summary-refactor

Reply:
Ah. Interesting. So that benchmark is saying that frugalVisitor is something like 4x faster in terms of seconds per operation? Definitely worth a refactor.

Reply:
Yes, I suspect the array case would be much bigger difference as well.

Reply:
Alright. With some changes to the internals of the MultibandTileFoldingVisitor, I was able to get the benchmark down from ~0.5s/op to the following:
```
[info] Benchmark                               Mode  Cnt  Score   Error  Units
[info] PolygonalSummaryBench.cleanVisitor      avgt    3  0.009 ± 0.026   s/op
[info] PolygonalSummaryBench.multibandVisitor  avgt    3  0.047 ± 0.014   s/op
```
These two benchmarks run on:
- raster/data/geotiff-test-files/overviews/singleband.tif
- raster/data/geotiff-test-files/overviews/multiband.tif

Which are the same RGBA four band tif, but the singleband one only has the R band. So the two tests are as equivalent as can be. We'd expect the multiband version to be slower and scale a bit with the number of bands read, so I think 0.009 vs 0.047 is about as good as we're gonna get for now.